docs: Draft a new Mir tutorial (#3496)
canonical/mir: doc/sphinx/make-your-own-compositor.md

## Using Mir Libraries to Build Your Own Wayland Compositor

> At this point, you're probably taken away by Mir's capabilities and want to use

Reviewer: "At this point" seems strange at the start of a tutorial

through the installation and basic usage of Mir. You'll be even more impressed
by how easy it is to create a Wayland compositor using Mir!

### A barebones Mir compositor

The following code block is the bare minimum you need to run a Mir compositor:
```cpp
#include <miral/runner.h>
#include <miral/minimal_window_manager.h>
#include <miral/set_window_management_policy.h>

using namespace miral;

int main(int argc, char const* argv[])
{
    MirRunner runner{argc, argv};

    return runner.run_with(
        {
            set_window_management_policy<MinimalWindowManager>()
        });
}
```

You can se it's barely longer than the usual C++ "Hello, world!" example. This
code should give you a minimal compositor with basic window management
capabilities such as controlling multiple windows, minimizing and maximizing,
handling input, etc... This is done with the help of MirAL (Mir Abstraction
Layer) which gives you a high level interface to work with Mir!

To compile, you'll need to install `libmiral`, and
`mir-graphics-drivers-desktop` which can be done on Ubuntu (and Debian
derivatives) with the following command:
```sh
sudo apt install libmiral-dev mir-graphics-drivers-desktop
```

> **Note**: `mir-graphics-drivers-desktop` is only needed for Debian based distros. 

Reviewer: Better a list for different platforms as in "explore"

<details>
<summary> Installing Mir Libraries on Alpine </summary>

```sh
sudo apk add mir
```
</details>

### A barebones Mir compositor

The following code block is the bare minimum you need to run a Mir compositor:
```cpp
#include <miral/runner.h>
#include <miral/minimal_window_manager.h>
#include <miral/set_window_management_policy.h>

using namespace miral;

int main(int argc, char const* argv[])
{
    MirRunner runner{argc, argv};

    return runner.run_with(
        {
            set_window_management_policy<MinimalWindowManager>()
        });
}
```

You can se it's barely longer than the usual C++ "Hello, world!" example. This
code should give you a minimal compositor with basic window management
capabilities such as controlling multiple windows, minimizing and maximizing,
handling input, etc... This is done with the help of MirAL (Mir Abstraction
Layer) which gives you a high level interface to work with Mir!

To compile, you'll need to install `libmircore`, `libmiral`, and
`mir-graphics-drivers-desktop` which can be done on Ubuntu (and Debian
derivatives) with the following command:
```sh
sudo apt install libmircore-dev libmiral-dev mir-graphics-drivers-desktop

> ```

Reviewer: 1. You don't need to specify libmircore-dev, it is a dependency of libmiral-dev
2. This conflicts with the "Installation" section above

**Note**: `mir-graphics-drivers-desktop` is only needed for Debian based distros. 

To compile this simple program:
```sh
g++ main.cpp -I/usr/include/mircore -I/usr/include/miral -lmiral -o demo-mir-compositor

> ```

Reviewer: Would it be kinder to provide a CMakeLists.txt that handles the dependencies and is more open to extension?

Author: I guess pkgconfig errors are a bit more readable than linker errors. Is there a way to check on `mir-graphics-drivers-desktop` though?

Reviewer: The graphics-drivers are a runtime dependency, so not simple to check until runtime

> ---

Reviewer 2: This file should be renamed to `tutorial.md` in my opinion

Author: Having descriptive names never hurts :)

Reviewer: Or `write-your-first-wayland-compositor.md`

discourse: 4911,5164,5603,6756,8037

> ---

Reviewer 2: What does this header do nowwadays?

Author: It creates this section in the sidebar:
![image](https://github.com/user-attachments/assets/15f463eb-0437-4991-b4d1-bdab37602cd9)

Reviewer: @mattkae you can follow the RTD link in the "checks" to see the "live" docs

> # Using Mir Libraries to Build Your Own Wayland Compositor

Reviewer 2: ```suggestion
# Tutorial
```

Reviewer 2: Even if we don't call this "Tutorial" it should be called "Build a Wayland Compositor". I think Mir is implied 😉

Reviewer: I agree this is verbose, but "Build" suggest checkout && make, we want something more along the lines of "Writing"...

"Creating A First Compositor"
"Writing A Simple Compositor"

Author: How does "Writing your first Wayland Compositor" sound?

If you've seen [the demo tutorial](explore-mir-using-demos), you're probably
taken away by Mir's capabilities and want to use it and the supporting
libraries to build your own Wayland compositor. This section will guide you
through the installation and basic usage of Mir. You'll be even more impressed
by how easy it is to create a Wayland compositor using Mir!

### A barebones Mir compositor

> The following code block is the bare minimum you need to run a Mir compositor:

Reviewer 2: I think that this is the wrong order of things. A tutorial should typically be linear in how its followed. I would recommend the following order:

1. Installation
2. Programming
3. Building
4. Running

Reviewer: Agreed

> The following code block is the bare minimum you need to run a Mir compositor:

Reviewer 2: We should also use some more "tutorial"-y language. For example, we could say "First, install Mir from the archive... Then, create a new file called `main.cpp` and write the following... Afterwards, create a new file called `CMakeLists.txt` and write the following... Finally, run these commands to build your project."

See https://diataxis.fr/tutorials/ for more details. In a nutshell, giving precise commands leaves no room for ambiguity

Author: I don't think the audience we're aiming for already knows the basics of C++. Making instructions more detailed might make things just noisier for readers

Reviewer 2: Yup I agree with that! But the more direct the tutorial is, the better in my opinion. Tutorials are suited for imperative language and a lot of "hand-holding", even when it may be obvious to an experienced user.

I always find that the best tutorials allow me to copy-and-paste most of it to get a minimum viable demo working and then I dive deeper into the completed tutorial afterward.

Author: I modified the tutorials a bit, what do you think about their current state?

```cpp
#include <miral/runner.h>
#include <miral/minimal_window_manager.h>
#include <miral/set_window_management_policy.h>

using namespace miral;

int main(int argc, char const* argv[])
{
    MirRunner runner{argc, argv};

    return runner.run_with(
        {
            set_window_management_policy<MinimalWindowManager>()
        });
}
```

> You can see it's barely longer than the usual C++ "Hello, world!" example. This
> code gives you a minimal compositor with basic window management
> capabilities such as controlling multiple windows, minimizing and maximizing,
> handling input, etc... This is done with the help of MirAL (Mir Abstraction
> Layer) which gives you a high level interface to work with Mir!

Reviewer 2: ```suggestion
This program creates a floating window manager with basic window management
capabilities such as controlling multiple windows, minimizing and maximizing,
and handling mouse input. This is done with the help of MirAL (Mir Abstraction
Layer) which gives you a high level interface to work with Mir.
```

I am a little biased against exclamation points by the way :wink: !

#### Dependencies
To compile, you'll need to install `libmiral`, and
`mir-graphics-drivers-desktop` which can be done on different distros as
follows:

Installing dependencies on Debian and its derivatives:
```sh
sudo apt install libmiral-dev mir-graphics-drivers-desktop
```

<details>
<summary> Installing Dependencies on Other Distros </summary>

Installing dependencies on Fedora 
```sh
sudo dnf install mir-devel libxkbcommon
```
Installing dependencies on Alpine 
```sh
sudo apk add mir-dev
```

> </details>

Reviewer 2: The tutorial can safely assume that you're on Ubuntu. This is because we can't guarantee anything about whether or not other distros are on latest Mir

Reviewer 2: We should mention our assumptions at the top of the tutorial, e.g.:

```md
## Prerequisites
- A computer running Ubuntu 24.04
- A compiler with support for C++23
```

Author: Hmmm, that would make things a lot easier to test (manually and automatically) and should allow us to ditch the collapsing sections needed for other distros. @AlanGriffiths What are your thoughts on this?

#### Building
To compile this simple program, you can use this `CMakeLists.txt` file:
```cmake
cmake_minimum_required(VERSION 3.5)

project(demo-mir-compositor)

> set(CMAKE_CXX_STANDARD 20)

Reviewer 2: We might want to make the standard be `23` :)

---
discourse: 4911,5164,5603,6756,8037
---

# Using Mir Libraries to Build Your Own Wayland Compositor
If you've seen [the demo tutorial](explore-mir-using-demos), you're probably
taken away by Mir's capabilities and want to use it and the supporting
libraries to build your own Wayland compositor. This section will guide you
through the installation and basic usage of Mir. You'll be even more impressed
by how easy it is to create a Wayland compositor using Mir!

### A barebones Mir compositor

The following code block is the bare minimum you need to run a Mir compositor:
```cpp
#include <miral/runner.h>
#include <miral/minimal_window_manager.h>
#include <miral/set_window_management_policy.h>

using namespace miral;

int main(int argc, char const* argv[])
{
    MirRunner runner{argc, argv};

    return runner.run_with(
        {
            set_window_management_policy<MinimalWindowManager>()
        });
}
```

> You can se it's barely longer than the usual C++ "Hello, world!" example. This

Reviewer: ```suggestion
You can see it's barely longer than the usual C++ "Hello, world!" example. This
```

> code should give you a minimal compositor with basic window management

Reviewer: ```suggestion
code gives you a minimal compositor with basic window management
```

capabilities such as controlling multiple windows, minimizing and maximizing,
handling input, etc... This is done with the help of MirAL (Mir Abstraction
Layer) which gives you a high level interface to work with Mir!

To compile, you'll need to install `libmiral`, and

> `mir-graphics-drivers-desktop` which is can be done on different distros as
> follows:

Reviewer: ```suggestion
`mir-graphics-drivers-desktop` which can be done on different distros as
follows:
```

<details>
<summary> Installing Dependencies on Debian and its derivatives </summary>

```sh
sudo apt install libmiral-dev mir-graphics-drivers-desktop
```

> </details>

Reviewer: Again, I don't think this "click to read on" experience is good.

Author: Having Ubuntu as the default and other distros in a collapsible section is a bit better, but I think it still could be better. No idea what can be changed though, perhaps the styling of the collapsible header itself?

Reviewer: Agreed. But unless there's a better suggestion I think this is good enough

other applications. For example you might want to start a background viewer, a
terminal, and a status bar. Doing so manually like what we've do would be
immensely annoying, so let's add an option for users to pass a list of startup
apps!

It's up to you how you want your command line interface will be. Let's say
we'll take a list of colon separated program names, then we'll launch each of
those individually. 

We'll start by creating an instance of `miral::ExternalClientLauncher`, this
will provide a way to launch programs by name. Then, we'll create a function
that will take the colon separated string and launch each application
separately. Finally, we just need to pass the external client launcher and our
configuration option to {func}`run_with` so that they're registered with the
compositor.

```diff
 #include <miral/runner.h>
+#include <miral/configuration_option.h>
+#include <miral/external_client.h>
 #include <miral/minimal_window_manager.h>
 #include <miral/set_window_management_policy.h>
 
@@ -9,8 +10,22 @@ int main(int argc, char const* argv[])
 {
     MirRunner runner{argc, argv};
 
+    miral::ExternalClientLauncher external_client_launcher;
+    

> +    auto run_startup_apps = [&](std::string const& apps) 
> +    {
> +        for(auto i = std::begin(apps); i != std::end(apps);)
> +        {
> +            auto const j = std::find(i, std::end(apps), ':');
> +            external_client_launcher.launch(std::vector<std::string>{std::string{i, j}});
> +            if((i = j) != std::end(apps)) ++i;
> +        }
> +    };

Reviewer: I know this comes from existing example code, but we can now add an option multiple times. (Maybe the answer could be updated.)

```suggestion
+    auto run_startup_apps = [&](std::vector<std::string> const& apps) 
+    {
+        for(auto const& app : apps)
+        {
+            external_client_launcher.launch(std::vector<std::string>{app}});
+        }
+    };
```

Author: One thing I just noticed while trying this out in code is that the compiler seems to not be able to infer that the lambda is convertible to `std::function<void(std::vector<std::string> const&)>`. If I replace `auto` with the type, it compiles.

Reviewer: Obviously, examples should compile. (We should find a way to automate that.)

I thought the `lambda_as_function<>` hack provided the necessary level of indirection. What was the error message?

Reviewer: @tarek-y-ismail see #3507

+
     return runner.run_with(
         {
-            set_window_management_policy<MinimalWindowManager>()
+            set_window_management_policy<MinimalWindowManager>(),
+            external_client_launcher,

> +            miral::ConfigurationOption{run_startup_apps, "startup-apps", "Colon separated list of startup apps", ""},

Reviewer: ```suggestion
+            miral::ConfigurationOption{run_startup_apps, "startup-app", "App to run at startup (can be specified multiple times)"},
```

         });
 }
```

Now, to start `kgx` and `bomber` on startup, we can do:
```sh

> ./build/demo-mir-compositor --startup-apps kgx:bomber

Reviewer: ```suggestion
./build/demo-mir-compositor --startup-app kgx --startup-app bomber
```

---
discourse: 4911,5164,5603,6756,8037
---

# Using Mir Libraries to Build Your Own Wayland Compositor

> If you've seen [the demo tutorial](explore-mir-using-demos), you're probably
> taken away by Mir's capabilities and want to use it and the supporting
> libraries to build your own Wayland compositor. This tutorial will guide you
> through the installation of dependencies, writing a simple compositor,
> building it, and running it. You'll be even more impressed by how easy it is to
> create a Wayland compositor using Mir!

Reviewer: I think we should lead with what the tutorial teaches.
```suggestion
This tutorial will guide you through writing a simple compositor: the installation of dependencies, building it, and running it. You'll be impressed by how easy it is to create a Wayland compositor using Mir!

As you continue to develop it your compositor has access to all the features of Mir demonstrated in [the demo tutorial](explore-mir-using-demos).
```

## Assumptions
This tutorial assumes that:
1. The reader is familiar with C++ and CMake.

> 2. The reader has a C++ compiler with C++ 23 support.

Reviewer: Mir/main needs C++23 support, but the tutorial doesn't and should "just work" with whatever compiler is on the reader's distro. (Because that's what that version of Mir was built with)

## A barebones Mir compositor
This section will cover the needed dependencies and how to install them, the
minimum code needed for a Mir compositor, how to build this code, and finally
how to run your compositor.

### Dependencies
Before you start coding, you'll need to install `libmiral`, and
`mir-graphics-drivers-desktop` which can be done on different distros as

> follows:

Reviewer: There are additional "build" dependencies: c++, cmake, etc.

Author: I think it's a bit too much to put that dependencies section. Would the assumptions section work? (see latest commit)

project(demo-mir-compositor)

set(CMAKE_CXX_STANDARD 23)

include(FindPkgConfig)
pkg_check_modules(MIRAL miral REQUIRED)
pkg_check_modules(XKBCOMMON xkbcommon REQUIRED)

add_executable(demo-mir-compositor main.cpp)

target_include_directories(demo-mir-compositor PUBLIC SYSTEM ${MIRAL_INCLUDE_DIRS})
target_link_libraries(     demo-mir-compositor               ${MIRAL_LDFLAGS})
target_link_libraries(     demo-mir-compositor               ${XKBCOMMON_LIBRARIES})
```

Then to build:
```sh
cmake -B build
cmake --build build
```

### Running
To run, you can run nested in an X or Wayland session, or from a virtual
terminal, just like the demo applications in [Explore What Mir Is Capable of
Using Mir Demos](explore-mir-using-demos.md). For example, if we were to run
inside an existing Wayland session, we'd run the following command:
```sh
WAYLAND_DISPLAY=wayland-99 ./build/demo-mir-compositor 
```
You'll see a window housing the compositor with a black void filling it. To
fill this void with some content, you can run the following from another
terminal:
```sh
WAYLAND_DISPLAY=wayland-99 bomber
```
You're free to replace `bomber` with any other Wayland compatible application.

> ## Startup Applications

Reviewer: I'm wondering now how many of these steps belong in "the" tutorial. There are many possibilities and we have to stop somewhere.

Maybe the tutorial ends here and each feature has a separate "HOWTO" starting from the above? Maybe there are a few "universal" features? Which?

Author: I think the tutorial should at end with a usable compositor, which would be the base + shortcuts

Reviewer: "Usable" is quite a flexible concept: for example, Frame is usable without shortcuts.

I'm OK with shortcuts, or even shortcuts and startup. Just flagging we need to draw a line.

@mattkae where would you put the line

Reviewer 2: I would draw the line at getting a simple window manager up and running. I think that the specifics of "How to do X in Mir" will each be a separate how-to that is linked at the bottom of the tutorial, e.g.:

```md
Now that you have your base compositor working, feel free to check out these guides on how to further develop your compositor:

- How to specify start up applications
- How to handle user input
- How to enable X11 applications
```

Reviewer: Unresolving as the review comments suggest changes that haven't happened

Author: The later sections were split off into their own how-tos, anything else missing?

+#include <miral/external_client.h>
 #include <miral/minimal_window_manager.h>
 #include <miral/set_window_management_policy.h>
 
@@ -9,8 +10,22 @@ int main(int argc, char const* argv[])
 {
     MirRunner runner{argc, argv};
 
+    miral::ExternalClientLauncher external_client_launcher;
+    
+    auto run_startup_apps = [&](std::vector<std::string> const& apps) 
+    {
+        for(auto const& app : apps)
+        {
+            external_client_launcher.launch(std::vector<std::string>{app}});
+        }
+    };
+
     return runner.run_with(
         {
-            set_window_management_policy<MinimalWindowManager>()
+            set_window_management_policy<MinimalWindowManager>(),
+            external_client_launcher,
+            miral::ConfigurationOption{run_startup_apps, "startup-app", "App to run at startup (can be specified multiple times)"},
         });
 }
```

Now, to start `kgx` and `bomber` on startup, we can do:
```sh
./build/demo-mir-compositor --startup-app kgx --startup-app bomber
```

## Keyboard shortcuts
What if you accidentally close the only open terminal? Or you want to log out?
Keyboard shortcuts are a good option that give you great flexibility without

> much coding effort. Let's see how they're implemented in Mir.

Reviewer: The last section described what was going to happen. This one should too.

It should say that we are going to make `Ctrl-Alt-T` launch a terminal and `Ctrl-AltBkSp` exit.
